pkg/cafs/cafs.go

```diff
@@ -252,3 +259,124 @@ func IsRootKey(fs storage.Store, key Key, leafSize uint32) bool {
 	}
 	return len(keys) > 0
 }
+
+func AddCaching(fsImpl Fs, cacheStore storage.Store, cafsKey2StoreKey func(Key) (string, error)) Fs {
+	return &cachedFs{
+		fsImpl:           fsImpl,
+		cacheStore:       cacheStore,
+		cafsKey2StoreKey: cafsKey2StoreKey,
+	}
+}
+
+type cachedFs struct {
+	fsImpl           Fs
+	cacheStore       storage.Store
+	cafsKey2StoreKey func(Key) (string, error)
+}
+
+func (fs *cachedFs) Put(ctx context.Context, src io.Reader) (int64, Key, []byte, bool, error) {
```

Comment:
Do we need to implement Put? Also, we should not be using ioutil.ReadAll interface.

Reply:
as of 17e09219ea37 , this falls back to the decorated `fsImpl`.